plugins/Lv2Effect/Lv2Effect.cpp

```diff
@@ -0,0 +1,113 @@
+/*
+ * Lv2Effect.cpp - implementation of LV2 effect
+ *
+ * Copyright (c) 2018-2019 Johannes Lorenz <j.git$$$lorenz-ho.me, $$$=@>
+ *
+ * This file is part of LMMS - https://lmms.io
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU General Public
+ * License as published by the Free Software Foundation; either
+ * version 2 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public
+ * License along with this program (see COPYING); if not, write to the
+ * Free Software Foundation, Inc., 51 Franklin Street, Fifth Floor,
+ * Boston, MA 02110-1301 USA.
+ *
+ */
+
+#include "Lv2Effect.h"
+
+#include <QDebug>
+#include <lv2.h>
+
+#include "Lv2SubPluginFeatures.h"
+
+#include "embed.h"
+#include "plugin_export.h"
+
+
+
+
+Plugin::Descriptor PLUGIN_EXPORT lv2effect_plugin_descriptor =
+{
+	STRINGIFY(PLUGIN_NAME),
+	"LV2",
+	QT_TRANSLATE_NOOP("pluginBrowser",
+		"plugin for using arbitrary LV2-effects inside LMMS."),
+	"Johannes Lorenz <j.git$$$lorenz-ho.me, $$$=@>",
+	0x0100,
+	Plugin::Effect,
+	new PluginPixmapLoader("logo"),
+	nullptr,
+	new Lv2SubPluginFeatures(Plugin::Effect)
+};
+
+
+
+
+Lv2Effect::Lv2Effect(Model* parent, const Descriptor::SubPluginFeatures::Key *key) :
+	Effect(&lv2effect_plugin_descriptor, parent, key),
+	m_controls(this, key->attributes["uri"])
+{
+}
+
+
+
+
+Lv2Effect::~Lv2Effect()
+{
+}
```

Comment:
I always thought you'll need a virtual destructor in every virtual class. Assumably though, since `Lv2Effect` inherits `Effect`, which does has a virtual destructor, `Lv2Effect`'s destructor is also virtual?

Reply:
At least, I get no warning when I remove this destructor...

Reply:
> since `Lv2Effect` inherits `Effect`, which does has a virtual destructor, `Lv2Effect`'s destructor is also virtual?

That's right - see [section 12.4.9 of the C++11 specification](https://timsong-cpp.github.io/cppwp/n3337/class.dtor#9). (Or [try it online](http://cpp.sh/3trtjy) - though the fact it works in one particular case doesn't always imply that it's defined behaviour.)

Reply:
Cool, how did you find that link?

I'll remove the destructor.

Reply:
> how did you find that link?

The official publications of the C++ specification are rather expensive, but the sources for the drafts are free on GitHub. Some has [a repository](https://github.com/timsong-cpp/cppwp#html-version-of-the-c-working-paper) where they provide compiled versions of them all. As for how to find the relevant bit of the specification, you get used to how it's organised. Destructors come under "special member functions" (which has been merged with the "classes" section in the latest draft).